### DISCUSSION TITLE
Find without sort key

### DISCUSSION THREAD

**Discussion by dominik-open-team** [sensedeep/dynamodb-onetable#133] [Q&A] [answered]
Hi guys! I'm learning this library and I'm having great time using it, but I encountered a problem.

This is my model: 

```
{
    PK: {
      type: [Function: String],
      value: '${_type}:credentialType:${credentialType}:value:${value}'
    },
    SK: { type: [Function: String], value: '${_type}' },
    value: { type: [Function: String], required: true },
    credentialType: { type: [Function: String], required: true },
    userId: { type: [Function: String], required: true },
    primary: { type: [Function: Boolean], default: false },
    GSI1pk: { type: [Function: String], value: 'userId:${userId}' },
    GSI1sk: {
      type: [Function: String],
      value: 'credentialType:${credentialType}:value:${value}'
    }
  }
```

And I would like to query every item with certain userId regardless of the values in sort key in GSI1. Unfortunetaly the library generates this dynamodb query 

```{
  "trace": {
    "cmd": {
      "ExpressionAttributeNames": {
        "#_0": "userId",
        "#_1": "GSI1pk",
        "#_2": "GSI1sk",
        "#_3": "_type"
      },
      "ExpressionAttributeValues": {
        ":_0": "user-4c6c048ec57a434cb44185a51ae674c5",
        ":_1": "userId:user-4c6c048ec57a434cb44185a51ae674c5",
        ":_2": "credentialType::value:",
        ":_3": "Credential"
      },
      "FilterExpression": "(#_0 = :_0) and (#_3 = :_3)",
      "KeyConditionExpression": "#_1 = :_1 and begins_with(#_2, :_2)",
      "TableName": "dominik-open-app-back-TheSingleTable",
      "ConsistentRead": false,
      "IndexName": "GSI1",
      "ScanIndexForward": true
    },
    "op": "find",
    "properties": {
      "userId": "user-4c6c048ec57a434cb44185a51ae674c5",
      "GSI1pk": "userId:user-4c6c048ec57a434cb44185a51ae674c5",
      "GSI1sk": {
        "begins": "credentialType::value:"
      },
      "_type": "Credential"
    },
    "params": {
      "parse": true,
      "high": true,
      "index": "GSI1",
      "checked": true
    }
  }
}
```
for this invocation 
```
await model.find({userId : userId},{ index: repoConstants.GSI1 })
```
and it returns no items because of this part of the KeyConditionExpression 

`begins_with(#_2, :_2)`

Am I doing something wrong, or its a bug?

**Comment by mobsense** [comment, marked as answer]
That looks correct to me.

Your GSI1 value template is:

    credentialType:${credentialType}:value:${value}

This means that all your GSI1 SK attributes start with "CredentialType:${credentialType}:value:"

When you don't provide a ${value}, then find will return all the matching items that being with:

    credentialType:${credentialType}:value:${value}

From the doc:

>If you call find or any query API and do not provide all the properties needed to resolve the complete value template. i.e. some of the ${var} references are unresolved, OneTable will take the resolved portion and create a begins with key condition for that portion of the value template.
